### PR TITLE
HDDS-5233. SCM subsequent init failed when previous scm init failed.

### PR DESCRIPTION
**Pull request by bharatviswa504**
…iled.

## What changes were proposed in this pull request?

Initialize version file before and then perform Ratis Server initialize. In this way, we shall use the same groupID in failure scenarios.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-5233
## How was this patch tested?

Existing CI.

### CODE REVIEW

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/StorageContainerManager.java [hunk 1]
```diff
@@ -946,19 +946,32 @@ public static boolean scmInit(OzoneConfiguration conf,
...
-          scmStorageConfig.setSCMHAFlag(true);
-        }
 
         if (OzoneSecurityUtil.isSecurityEnabled(conf)) {
           HASecurityUtils.initializeSecurity(scmStorageConfig, conf,
               getScmAddress(haDetails, conf), true);
         }
+
+        // Do not move these code lines. If SCM version file is created,
+        // the subsequent scm init should use the groupID from version file.
```
Comment (bshashikant) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/StorageContainerManager.java:956]: Do not move these code lines -> ensure scmRatisserver#initialize() is called post scm storage config initialization.

Comment (bshashikant) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/StorageContainerManager.java:956]: groupId-> clusterId

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/StorageContainerManager.java [hunk 2]
```diff
@@ -946,19 +946,32 @@ public static boolean scmInit(OzoneConfiguration conf,
...
+        // So, scmStorageConfig#initialize() should happen before ratis server
+        // initialize. In this way,we do not leave ratis storage directory
+        // with multiple raft Groups in failure scenario.
+
```
Comment (bshashikant) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/StorageContainerManager.java:960]: multiple raft Groups -> multiple raft group directories

**Review by bshashikant — COMMENTED** [review]
@bharatviswa504 , can you also fix the SCM stateMachine registry code as well to ensure SCMStateMachine instance is not shared across multiple raftServerImpl instances.

**Comment by bharatviswa504** [comment]
> @bharatviswa504 , can you also fix the SCM stateMachine registry code as well to ensure SCMStateMachine instance is not shared across multiple raftServerImpl instances.

Updated code to fix this.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMStateMachine.java [hunk 3]
```diff
@@ -84,20 +81,16 @@
...
     TransactionInfo latestTrxInfo = this.transactionBuffer.getLatestTrxInfo();
-    if (!latestTrxInfo.isDefault() &&
-        !updateLastAppliedTermIndex(latestTrxInfo.getTerm(),
-            latestTrxInfo.getTransactionIndex())) {
-      throw new SCMException(
-          String.format("Failed to update LastAppliedTermIndex " +
-                  "in StateMachine to term:{} index:{}",
-              latestTrxInfo.getTerm(), latestTrxInfo.getTransactionIndex()
-          ), SCM_NOT_INITIALIZED);
+    if (!latestTrxInfo.isDefault()) {
```
[the remark this answers is not in the document]
Comment (GlenGeng) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMStateMachine.java:89]: Overall I am fine with the change.
Correctly me if I am wrong, the two are the same, except not checking the return code of `updateLastAppliedTermIndex()` ?

[the remark this answers is not in the document]
Comment (bharatviswa504) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMStateMachine.java:89]: Yes, previously updateLastAppliedTermIndex returns a false case we used to throw exception, which is not required from my understanding.

[the remark this answers is not in the document]
Comment (GlenGeng) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMStateMachine.java:89]: agree. It should never fail during the startup of StateMachine.

**Review by bshashikant — APPROVED** [review]

**Comment by bharatviswa504** [comment]
Thank You @bshashikant and @GlenGeng for the review

**Pull request merged**